### DISCUSSION TITLE
What does exactly AddHint method do?

### DISCUSSION THREAD

**Discussion by CADBIMDeveloper** [google/or-tools#3277] [CP-SAT questions]
Hi!

I'm trying to feed solver with start solution. I add hints for each variables (from definitely feasible solution). However it seems the solver doesn't take it into account.

My feasible solution objective value is 56
I added solution callbacks and print the time and objective values for intermediate solutions and this is what I see:
[00:00:02.6110171] Objective: 135
[00:00:02.8427712] Objective: 67
[00:00:02.9632210] Objective: 59
[00:00:03.1517073] Objective: 58
[00:00:05.0168057] Objective: 57
[00:00:35.4816558] Objective: 56 <- after 35 seconds solver gets the solution with the same objective value
[00:00:50.5899486] Objective: 55
...

**Comment by lperron** [comment]
you should enable log_search_progress.

**Comment by CADBIMDeveloper** [comment]
Hi Laurent!

Okay, I added:
```cs
solver.StringParameters = $"num_search_workers:{workers}{(enableLogging ? ";log_search_progress:true" : string.Empty)};max_memory_in_mb:1024;catch_sigint_signal:false{(workers == 1 ? ";random_seed:71272241" : string.Empty)}";

if (enableLogging)
      solver.SetLogCallback(message => File.AppendAllText(@"D:\Projects\00Temp\solverLog.txt", $"{message}\n"));
```

And now I have:

Starting CP-SAT solver v9.2.9972
Parameters: max_memory_in_mb: 1024 log_search_progress: true num_search_workers: 8 catch_sigint_signal: false

Initial optimization model '':
#Variables: 14625 (585 in objective)
  - 14625 in [0,1]
#kLinearN: 942 (#terms: 42705)

Starting presolve at 0.01s
The solution hint is complete and is feasible.
[Probing] deterministic_time: 0.0789457 (limit: 1) wall_time: 0.136425 (14625/14625)
[Probing]  - new binary clause: 14040
[Symmetry] Graph for symmetry has 16152 nodes and 43290 arcs.
[Symmetry] GraphSymmetryFinder error: Some automorphisms were found, but probably not all.
[Symmetry] Symmetry computation done. time: 0.720085 dtime: 1.00002
[Symmetry] #generators: 255, average support size: 250
[Symmetry] 195 orbits with sizes: 9072,432,378,294,216,42,42,42,42,42,...
[Symmetry] Num fixable by binary propagation in orbit: 41 / 9072
[Symmetry] Num fixable by intersecting at_most_one with orbits: 42 largest_orbit: 9072
[Symmetry] Found orbitope of size 45 x 2

Presolve summary:
  - 0 affine relations were detected.
  - 0 variable equivalence relations were detected.
  - rule 'exactly_one: removed literals' was applied 1 time.
  - rule 'linear: fixed or dup variables' was applied 84 times.
  - rule 'linear: positive equal one' was applied 312 times.
  - rule 'linear: simplified rhs' was applied 585 times.
  - rule 'presolve: 42 unused variables removed.' was applied 1 time.
  - rule 'presolve: iteration' was applied 1 time.
  - rule 'symmetry: fixed to false in general orbit' was applied 42 times.

Presolved optimization model '':
#Variables: 14583 (585 in objective)
  - 14583 in [0,1]
#kExactlyOne: 312 (#literals: 13998)
#kLinearN: 630 (#terms: 28581)

Preloading model.
[Symmetry] Graph for symmetry has 16110 nodes and 43164 arcs.
[Symmetry] GraphSymmetryFinder error: Some automorphisms were found, but probably not all.
[Symmetry] Symmetry computation done. time: 0.717447 dtime: 1.00007
[Symmetry] #generators: 264, average support size: 240
[Symmetry] Found orbitope of size 45 x 2
#Bound   2.01s best:inf   next:[0,585]    initial_domain
#Model   2.04s var:14583/14583 constraints:942/942

Starting Search at 2.04s with 8 workers and subsolvers: [ default_lp, reduced_costs, pseudo_costs, no_lp, max_lp, core, feasibility_pump, rnd_var_lns_default, rnd_cst_lns_default, graph_var_lns_default, graph_cst_lns_default, rins_lns_default, rens_lns_default ]
#1       2.43s best:135   next:[0,134]    rens_lns_default_incomplete(d=0.50 s=8 t=0.10 p=0.00)
#Bound   2.60s best:135   next:[45,134]   core
#2       2.67s best:56    next:[45,55]    no_lp [hint] fixed_bools:0/14583

Sub-solver search statistics:
  'reduced_costs':
     LP statistics:
       final dimension: 896 rows, 14583 columns, 33288 entries with magnitude in [3.779645e-01, 1.000000e+00]
       total number of simplex iterations: 242042
       num solves: 
         - #OPTIMAL: 2496
         - #DUAL_UNBOUNDED: 885
         - #DUAL_FEASIBLE: 9743
       managed constraints: 5408
       merged constraints: 4835
       total cuts added: 4466 (out of 9301 calls)
         - 'clique': 4466

  'pseudo_costs':
     LP statistics:
       final dimension: 2638 rows, 14583 columns, 45468 entries with magnitude in [3.779645e-01, 1.000000e+00]
       total number of simplex iterations: 131190
       num solves: 
         - #OPTIMAL: 1899
         - #DUAL_UNBOUNDED: 8
         - #DUAL_FEASIBLE: 1927
       managed constraints: 7464
       merged constraints: 4628
       total cuts added: 6522 (out of 11150 calls)
         - 'clique': 6522

  'max_lp':
     LP statistics:
       final dimension: 2734 rows, 14583 columns, 45799 entries with magnitude in [3.779645e-01, 1.000000e+00]
       total number of simplex iterations: 111329
       num solves: 
         - #OPTIMAL: 2818
         - #DUAL_UNBOUNDED: 6
         - #DUAL_FEASIBLE: 594
       managed constraints: 8286
       merged constraints: 6365
       total cuts added: 7344 (out of 13709 calls)
         - 'clique': 7344


Solutions found per subsolver:
  'no_lp': 1
  'rens_lns_default_incomplete': 1

Objective bounds found per subsolver:
  'core': 1
  'initial_domain': 1

CpSolverResponse summary:
status: FEASIBLE
objective: 56
best_bound: 45
booleans: 14583
conflicts: 348
branches: 48754
propagations: 791795
integer_propagations: 916836
restarts: 29169
lp_iterations: 111329
walltime: 121.532
usertime: 121.532
deterministic_time: 331.548
gap_integral: 795.345

I see "The solution hint is complete and is feasible" but the hinted solution comes much later...

**Comment by lperron** [comment]
So one worker follows the hint, and another one is faster to find a
solution without the hint.

--Laurent

PS: You should switch to 9.3
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by CADBIMDeveloper** [comment]
Sounds logical, but 
1) Why it takes so much time to get feasible solution I provided as a hint?
[00:00:35.4816558] Objective: 56 <- after 35 seconds solver gets the solution with the same objective value
2) I tried to set exactly 1 worker:
Parameters: random_seed: 71272241 max_memory_in_mb: 1024 log_search_progress: true num_search_workers: 1 catch_sigint_signal: false

And I saw this:
[00:00:13.3482490] Objective: 59
[00:00:23.2358324] Objective: 56 <- the same objective value as in hint
[00:00:27.8166272] Objective: 55
...

**Comment by lperron** [comment]
do you have a search strategy ?

Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by CADBIMDeveloper** [comment]
Do you mean CpModel.AddDecisionStrategy? No...

**Comment by lperron** [comment]
Is your model big?

**Comment by CADBIMDeveloper** [comment]
In terms of variables/constraints count I guess it's not too big. It has ~14500 variables, all of them are booleans. It's an assignment-like problem, In general only few "instances" could be placed to a specific row, but I'm playing  with worst-case scenario where each instance could be placed to any row. I'm currently trying to figure out how to add symmetry-breaking constraints, it seems search space is extremely big...

**Comment by lperron** [comment]
can you send me the proto ?
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by CADBIMDeveloper** [comment]
Yes, of course
[model.zip](https://github.com/google/or-tools/files/8629713/model.zip)

I guess, you should set timeout or interrupt search in 1-2 minutes. 

One more thing to consider, I use 9.2, because it's the latest version available for .Net 4.8 (it's a part of CAD plugin, and I'm not able to use .Net 5 / 6 :-( ). Anyway if it works on the latest build as expected, just let me know :-) Thank you!

**Comment by lperron** [comment]
I cannot parse the model. I forgot if the format changed between 9.2 and 9.3

**Comment by CADBIMDeveloper** [comment]
I don't know what is the difference between 
`model.ExportToFile(...)`
and 
`model.Model.WriteTo(new CodedOutputStream(...`

But result files are slightly different

This one is from model.Model.WriteTo

[model1.zip](https://github.com/google/or-tools/files/8629950/model1.zip)

**Comment by lperron** [comment]
Not yet :-)

Can you try

model.ExportToFile("name_ending_with.txt");

in C#, model write uses a json format. ExportToFile uses the C++ exporter,
which uses a text format instead.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by CADBIMDeveloper** [comment]
Sure:

[model.txt](https://github.com/google/or-tools/files/8630180/model.txt)

**Comment by lperron** [comment]
```
I0505 11:51:32.288879     1 sat_runner.cc:176] Reading a CpModelProto.

Starting CP-SAT solver v9999.0.0
Parameters: log_search_progress: true num_search_workers: 1

Initial satisfaction model '':
#Variables: 14625
  - 14625 in [0,1]
#kLinearN: 942 (#terms: 42705)

Starting presolve at 0.00s
The solution hint is complete and is feasible.
[ExtractEncodingFromLinear] #potential_supersets=312 #potential_subsets=0 #at_most_one_encodings=0 #exactly_one_encodings=0 #unique_terms=0 #multiple_terms=0 #literals=0 time=0.000157s
[Probing] deterministic_time: 0.0730548 (limit: 1) wall_time: 0.079527 (14625/14625)
[DetectDuplicateConstraints] #duplicates=0 time=0.00059s
[DetectDominatedLinearConstraints] #relevant_constraints=630 #work_done=294840 #num_inclusions=585 #num_redundant=585 time=0.010301s
[DetectOverlappingColumns] #processed_columns=0 #work_done=0 #nz_reduction=0 time=0.000535s
[ProcessSetPPC] #relevant_constraints=357 #num_inclusions=0 work=84240 time=0.001347s
[Symmetry] Graph for symmetry has 14982 nodes and 28080 arcs.
[Symmetry] GraphSymmetryFinder error: Some automorphisms were found, but probably not all.
[Symmetry] Symmetry computation done. time: 0.455754 dtime: 1.00001
[Symmetry] #generators: 352, average support size: 153.716
[Symmetry] 5 orbits with sizes: 13062,622,311,42,2
[Symmetry] Num fixable by binary propagation in orbit: 41 / 13062
[Symmetry] Num fixable by intersecting at_most_one with orbits: 42 largest_orbit: 13062
[Symmetry] Found orbitope of size 45 x 2

Presolve summary:
  - 0 affine relations were detected.
  - rule 'exactly_one: removed literals' was applied 1 time.
  - rule 'linear inclusion: redundant included constraint' was applied 585 times.
  - rule 'linear: fixed or dup variables' was applied 42 times.
  - rule 'linear: positive equal one' was applied 312 times.
  - rule 'linear: simplified rhs' was applied 1170 times.
  - rule 'linear: singleton column' was applied 585 times.
  - rule 'presolve: 42 unused variables removed.' was applied 1 time.
  - rule 'presolve: iteration' was applied 1 time.
  - rule 'symmetry: fixed to false in general orbit' was applied 42 times.

Presolved satisfaction model '':
#Variables: 13998
  - 13998 in [0,1]
#kExactlyOne: 312 (#literals: 13998)
#kLinearN: 45 (#terms: 13998)

Preloading model.
[Symmetry] Graph for symmetry has 14355 nodes and 27996 arcs.
[Symmetry] Symmetry computation done. time: 0.44606 dtime: 0.950861
[Symmetry] #generators: 351, average support size: 152.148
[Symmetry] Found orbitope of size 45 x 2

Starting to load the model at 1.13s
Added 351 symmetry to the SAT solver.
[Probing] deterministic_time: 0.0724489 (limit: 1) wall_time: 0.172953 (13998/13998)
[LinearRelaxationBeforeCliqueExpansion] #linear:357 #at_most_ones:0
[FinalLinearRelaxation] #linear:357 #cut_generators:0
Initial num_bool: 13998

Starting sequential search at 1.49s
#1       2.71s  [hint] fixed_bools:0/13998

LP statistics:
  final dimension: 351 rows, 13998 columns, 27726 entries with magnitude in [1.000000e+00, 1.000000e+00]
  total number of simplex iterations: 10020
  num solves: 
    - #OPTIMAL: 515
    - #DUAL_FEASIBLE: 3
  managed constraints: 357
  total cuts added: 0 (out of 0 calls)


CpSolverResponse summary:
status: OPTIMAL
objective: NA
best_bound: NA
booleans: 0
conflicts: 0
branches: 0
propagations: 0
integer_propagations: 0
restarts: 0
lp_iterations: 0
walltime: 2.70874
usertime: 2.70874
deterministic_time: 0.752632
gap_integral: 0
```

I guess it works on master.

**Comment by CADBIMDeveloper** [comment]
Ahh, my mistake, it's before objective was set. I'll attach an updated model, sorry

**Comment by CADBIMDeveloper** [comment]
The model with an objective:

[model.txt](https://github.com/google/or-tools/files/8630247/model.txt)

**Comment by lperron** [comment]
1 worker: (stopped)

```
Starting sequential search at 1.70s
#1       9.73s best:59    next:[0,58]      [hint] fixed_bools:0/14583
#Bound   9.92s best:59    next:[7,58]     
#Bound  10.08s best:59    next:[16,58]    
#Bound  10.26s best:59    next:[19,58]    
#Bound  10.68s best:59    next:[27,58]    
#Bound  11.50s best:59    next:[28,58]    
#Bound  11.94s best:59    next:[29,58]    
#Bound  13.02s best:59    next:[34,58]    
#Bound  13.53s best:59    next:[35,58]    
#Bound  14.25s best:59    next:[39,58]    
#Bound  15.59s best:59    next:[44,58]    
#Bound  16.22s best:59    next:[45,58]    
#2      19.80s best:55    next:[45,54]     fixed_bools:0/14583
#3      24.69s best:54    next:[45,53]     fixed_bools:0/14583
```

16 workers:
```
Starting Search at 1.29s with 16 workers.
10 full subsolvers: [default_lp, no_lp, max_lp, core, reduced_costs, pseudo_costs, quick_restart, quick_restart_no_lp, lb_tree_search, probing]
Interleaved subsolvers: [feasibility_pump, rnd_var_lns_default, rnd_cst_lns_default, graph_var_lns_default, graph_cst_lns_default, rins_lns_default, rens_lns_default]
#1       1.56s best:115   next:[0,114]    rens_lns_default_incomplete(d=0.50 s=12 t=0.10 p=0.00)
#Bound   1.66s best:115   next:[1,114]    bool_core num_cores:1 [core:3 mw:1 d:0] assumptions:583 depth:2 fixed_bools:0/14586
#Bound   1.67s best:115   next:[2,114]    bool_core num_cores:2 [core:44 mw:1 d:2] assumptions:540 depth:6 fixed_bools:0/14671
#Bound   1.67s best:115   next:[3,114]    bool_core num_cores:3 [core:44 mw:1 d:6] assumptions:497 depth:6 fixed_bools:0/14756
#Bound   1.68s best:115   next:[4,114]    bool_core num_cores:4 [core:44 mw:1 d:6] assumptions:454 depth:6 fixed_bools:0/14841
#Bound   1.69s best:115   next:[5,114]    bool_core num_cores:5 [core:44 mw:1 d:6] assumptions:411 depth:6 fixed_bools:0/14926
#Bound   1.70s best:115   next:[6,114]    bool_core num_cores:6 [core:44 mw:1 d:6] assumptions:368 depth:6 fixed_bools:0/15011
#Bound   1.71s best:115   next:[7,114]    bool_core num_cores:7 [core:44 mw:1 d:6] assumptions:325 depth:6 fixed_bools:0/15096
#Bound   1.72s best:115   next:[8,114]    bool_core num_cores:8 [core:44 mw:1 d:6] assumptions:282 depth:6 fixed_bools:0/15181
#Bound   1.72s best:115   next:[9,114]    bool_core num_cores:9 [core:44 mw:1 d:6] assumptions:239 depth:6 fixed_bools:0/15266
#Bound   1.73s best:115   next:[10,114]   bool_core num_cores:10 [core:44 mw:1 d:6] assumptions:196 depth:6 fixed_bools:0/15351
#Bound   1.73s best:115   next:[11,114]   bool_core num_cores:11 [core:44 mw:1 d:6] assumptions:153 depth:6 fixed_bools:0/15436
#Bound   1.74s best:115   next:[12,114]   bool_core num_cores:12 [core:44 mw:1 d:6] assumptions:110 depth:6 fixed_bools:0/15521
#Bound   1.75s best:115   next:[13,114]   bool_core num_cores:13 [core:44 mw:1 d:6] assumptions:67 depth:6 fixed_bools:0/15606
#Bound   1.75s best:115   next:[14,114]   bool_core num_cores:14 [core:13 mw:1 d:6] assumptions:55 depth:6 fixed_bools:0/15629
#2       1.95s best:111   next:[14,110]   graph_var_lns_default(d=0.50 s=20 t=0.10 p=0.00)
#3       2.05s best:59    next:[14,58]    no_lp [hint] fixed_bools:0/14583
#4       2.17s best:58    next:[14,57]    no_lp fixed_bools:0/14583
#5       4.26s best:57    next:[14,56]    graph_var_lns_default(d=0.79 s=55 t=0.10 p=0.75)
#Bound   5.11s best:57    next:[28,56]    max_lp
#6       5.73s best:56    next:[28,55]    quick_restart_no_lp fixed_bools:0/14583
#Bound   6.66s best:56    next:[45,55]    max_lp
```
Anyway, hinting works as expected. Just remember that not all workers try to follow hints.

**Comment by CADBIMDeveloper** [comment]
That's good, thank you! I'm just curious, why if we have only one worker, it doesn't follow the hints?

**Comment by lperron** [comment]
It does. First solution is marked with [hint].

**Comment by CADBIMDeveloper** [comment]
Hi Laurent.

These two models are the same except hints. Hinted solution objective value is 56 in both cases.

On both models solver writes:
`The solution hint is complete and is feasible.`

[solver-model_56.txt](https://github.com/google/or-tools/files/8638735/solver-model_56.txt)
[solver-model_59.txt](https://github.com/google/or-tools/files/8638737/solver-model_59.txt)

In the first case the solution is exactly equal to hinted

On the second case these values are different:
```
- assignment_3_0 = 1 (hint=0)
- assignment_3_203 = 0 (hint=1)
- portion_3_0_usage = 1 (hint=0)
- assignment_29_0 = 0 (hint=1)
- assignment_29_311 = 1 (hint=0)
- portion_29_12_usage = 1 (hint=0)
- assignment_33_297 = 1 (hint=0)
- assignment_33_311 = 0 (hint=1)
- assignment_34_287 = 1 (hint=0)
- assignment_34_297 = 0 (hint=1)
- assignment_36_279 = 1 (hint=0)
- assignment_36_287 = 0 (hint=1)
- assignment_37_265 = 1 (hint=0)
- assignment_37_279 = 0 (hint=1)
- assignment_38_263 = 1 (hint=0)
- assignment_38_265 = 0 (hint=1)
- assignment_39_261 = 1 (hint=0)
- assignment_39_263 = 0 (hint=1)
- assignment_40_247 = 1 (hint=0)
- assignment_40_261 = 0 (hint=1)
- assignment_41_239 = 1 (hint=0)
- assignment_41_247 = 0 (hint=1)
- assignment_43_229 = 1 (hint=0)
- assignment_43_239 = 0 (hint=1)
- assignment_44_203 = 1 (hint=0)
- assignment_44_229 = 0 (hint=1)
- portion_44_8_usage = 1 (hint=0)
```

A part of solver log from the first model:
```
#1       2.41s best:135   next:[0,134]    rens_lns_default_incomplete(d=0.50 s=9 t=0.10 p=0.00)
#Bound   2.51s best:135   next:[45,134]   core
#2       2.62s best:67    next:[45,66]    core fixed_bools:0/14584
#3       2.65s best:56    next:[45,55]    no_lp [hint] fixed_bools:0/14583
```

A part of solver log from the second model:
```
#1       2.57s best:135   next:[0,134]    rens_lns_default_incomplete(d=0.50 s=8 t=0.10 p=0.00)
#Bound   2.65s best:135   next:[45,134]   core
#2       2.77s best:67    next:[45,66]    core fixed_bools:0/14584
#3       2.89s best:59    next:[45,58]    no_lp [hint] fixed_bools:0/14583
```

**Reply by lperron** [reply]
are you sure that both models are hinted at a value of 56 ? the hints are vastly different, and the first solutions are found without any conflict.

**Comment by CADBIMDeveloper** [comment]
Yes, I'm totally sure.

Let me briefly describe the model. It has a matrix of boolean "assignment_(i)_(j)" variables. 
In each row we must have exact values of assignments, e.g.
`sum[assignment_row_0...assignment_row_311] = row_constraint`
Each instance can be assigned exactly one time, e.g.
`sum[assignment_0_column...assignment_44_column = 1`

Instances are grouped into portions. There are 13 portions, each one contains exactly 24 instances. For each row in assignment matrix I add 13 variables
`[portion_row_0_usage, portion_row_1_usage, ..., portion_row_12_usage]`

and constraints for them:
`sum[assignment_row_0...assignment_row_23] <= row_value_constraint*portion_row_0_usage`
`sum[assignment_row_24...assignment_row_47] <= row_value_constraint*portion_row_1_usage;`
...
`sum[assignment_row_288...assignment_row_311] <= row_value_constraint*portion_row_12_usage;`

The goal is to minimize mixing portions in each row, ideally we have instances of a single portion in each row.

So the model objective is a sum of all portion usages variable.

The full story is a bit more complex, this model comes from a bit synthetic input data. 

So. let's back to your question. Solver consider hinted solution as complete and feasible (and it is complete and feasible by design), I'm adding hints for each variable, so I add zero or one to the basic objective value each time I add hints for portion usages variables. That's why I'm sure that hinted solution objective value is equal to 56.

I've already attached the differences between hinted and calculated solution, portion_3_0_usage, portion_29_12_usage and portion_44_8_usage in calculated solution are ones instead of hinted zeroes. Thats why we have 59 instead of 56.

Some insights are also come from this differences log:

- in each row two values are swapped, e.g. assignment_3_0 = 1 instead of 0 and assignment_3_203 = 0 instead of 1. Other values are the same, so all rows constraints are satisfied for both solutions. 
- all columns constraints are also satisfied, e.g. assignment_3_0 = 1 instead of 0 and assignment_29_0 = 0 instead of1. 

Portions usage constraints are also ok for both solutions, they can't strict assignments variables in any way...
